tools/shared/src/main/scala/org/scalajs/core/tools/linker/analyzer/Analyzer.scala

```diff
@@ -196,48 +246,40 @@ private final class Analyzer(config: CommonPhaseConfig,
     val isExported = data.isExported
 
     var superClass: Option[ClassInfo] = _
+    var interfaces: List[ClassInfo] = _
     var ancestors: List[ClassInfo] = _
-    val descendants = mutable.ListBuffer.empty[ClassInfo]
-
-    var nonExistent: Boolean = false
 
-    /** Ensures that this class and its dependencies are linked.
-     *
-     *  @throws CyclicDependencyException if this class is already linking
-     */
-    def linkClasses(): Unit = {
-      if (_linking)
-        throw CyclicDependencyException(this :: Nil)
-
-      if (!_linked) {
-        _linking = true
-        try {
-          linkClassesImpl()
-        } catch {
-          case CyclicDependencyException(chain) =>
-            throw CyclicDependencyException(this :: chain)
-        }
-        _linking = false
-        _linked = true
+    /** Links this class to its superclass and implemented interfaces. */
+    def linkClasses()(implicit from: From): Unit = {
+      assert(!linked)
+      linkedFrom ::= from
+      try {
+        linkClassesImpl()
+      } catch {
+        case CyclicDependencyException(chain, _) =>
+          throw CyclicDependencyException(this :: chain, from)
       }
+      linked = true
     }
 
     private[this] def linkClassesImpl(): Unit = {
-      superClass = data.superClass.map(lookupClass(_))
+      implicit val from = FromClassHeader(this)
 
-      val parents = data.superClass ++: data.interfaces
+      def lookupClassAndCheckCycles(encodedName: String): ClassInfo = {
+        val info = lookupClass(encodedName)
+        if (!info.linked)
+          throw CyclicDependencyException(Nil, from)
+        info
+      }
 
-      ancestors = this +: parents.flatMap { parent =>
-        val cls = lookupClass(parent)
-        cls.linkClasses()
-        cls.ancestors
-      }.distinct
+      superClass = data.superClass.map(lookupClassAndCheckCycles(_))
+      interfaces = data.interfaces.map(lookupClassAndCheckCycles(_))
 
-      for (ancestor <- ancestors)
-        ancestor.descendants += this
-    }
+      // TODO #3076: Validate the super class and implemented interfaces
```

Comment:
The current PR does not improve nor worsen the situation wrt. #3076. Addressing that issue should be done in a separate PR. The new architecture of the Analyzer will make it easier, though, as we get here only for classes that really need to be linked (needed at all), which means we can eagerly report errors if things don't match. With the previous architecture, this was virtually impossible to do.